### PR TITLE
Make GenerateEventsFilter work with all MatrixWorkspace.

### LINKED ISSUE #25542: Enable GenerateEventsFilter to work with Workspace2D
### Expected behavior

GenerateEventsFilter shall work with any MatrixWorkspace that has run object and sample logs in form of TimeSeriesProperty.

### Actual behavior

It works for EventWorkspace only.

### Steps to reproduce the behavior

1. Create an empty Workspace2D
2. Execute LoadNexusLogs
3. Execute GenerateEventsFilter

### PR DESCRIPTION
**Pull request by wdzhou**
**Description of work.**

Enable GeneratesEventFilter to work with any MatrixWorkspace containing **run** and sample logs.

**To test:**

Execute the following script from both master and this branch and compare.

```
empty = CreateWorkspace(DataX=[0], DataY=[0], DataE=[1], NSpec=1)
empty = LoadNexusLogs(Workspace=empty, Filename='/SNS/VULCAN/IPTS-22862/nexus/VULCAN_172282.nxs.h5')
GenerateEventsFilter(InputWorkspace='empty', OutputWorkspace='MatrixSlicerss', InformationWorkspace='MatrixInfoTabless', TimeInterval='300', LogName='loadframe.furnace2', MinimumLogValue=100, MaximumLogValue=1000, LogValueInterval=100)
GenerateEventsFilter(InputWorkspace='empty', OutputWorkspace='MatrixSlicer', InformationWorkspace='MatrixInfoTable', TimeInterval='300')

```

Fixes #25542. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
This is a good improvement to `GenerateEventsFilter`. A couple of small changes:
- Add release notes and/or update the algorithm's documentation. I think the example in the PR testing instructions would be really useful for users.
- The exceptions generated would benefit from greater clarity. I think having the workspace name and the fact that the `proton_charge` log is missing is useful in both (current error doesn't give the log's proper name). Then the extra "not an event workspace" or "is empty event workspace." Hopefully users never see the errors, but when they do they'll benefit from the clarity.

**Comment by wdzhou** [comment]
@peterfpeterson all of your suggestion have been implemented.

File: Framework/Algorithms/src/GenerateEventsFilter.cpp [hunk 1]
```diff
@@ -1882,12 +1885,36 @@ DateAndTime GenerateEventsFilter::findRunEnd() {
...
-    for (size_t i = 0; i < m_dataWS->getNumberHistograms(); ++i) {
-      const DataObjects::EventList &evlist = m_dataWS->getSpectrum(i);
+    runendtime = 0;
+
+    DataObjects::EventWorkspace_const_sptr eventWS =
+        boost::dynamic_pointer_cast<const EventWorkspace>(m_dataWS);
+    if (!eventWS) {
+      stringstream errss;
+      errss << "Input workspace " << m_dataWS->getName()
+            << " is not an Eventworkspace and does not have sample 'log "
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/GenerateEventsFilter.cpp]:
```suggestion
            << " is not an Eventworkspace and does not have sample log "
```

File: Framework/Algorithms/src/GenerateEventsFilter.cpp [hunk 2]
```diff
@@ -1882,12 +1885,36 @@ DateAndTime GenerateEventsFilter::findRunEnd() {
...
+               "proton_charge'."
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/GenerateEventsFilter.cpp]:
```suggestion
               "'proton_charge'."
```

File: Framework/Algorithms/src/GenerateEventsFilter.cpp [hunk 3]
```diff
@@ -1882,12 +1885,36 @@ DateAndTime GenerateEventsFilter::findRunEnd() {
...
+               "Therefore it fails to find run end time.";
+      g_log.error(errss.str());
+
+      throw std::runtime_error(errss.str());
+    } else if (eventWS->getNumberEvents() == 0) {
+      stringstream errss;
+      errss << "Input EventWorkspace " << m_dataWS->getName()
+            << " has zero event and does not have sample log 'proton_charge'.  "
+               "Therefore, o Unable to "
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/GenerateEventsFilter.cpp]:
```suggestion
               "Therefore, unable to "
```

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Some typos then it is good

File: Framework/Algorithms/src/GenerateEventsFilter.cpp (hunk 3 above)
Comment (wdzhou) [reply Framework/Algorithms/src/GenerateEventsFilter.cpp]: where did this 'o' come from?

File: Framework/Algorithms/src/GenerateEventsFilter.cpp (hunk 2 above)
Comment (wdzhou) [reply Framework/Algorithms/src/GenerateEventsFilter.cpp]: ' was in a bad place.

**Review by peterfpeterson — APPROVED** [review]

**Pull request merged**